### PR TITLE
[w32process] Reduce mmap allocation by mapping file only once

### PR DESCRIPTION
**Pull request by luhenry**

### CODE REVIEW

File: mono/metadata/w32process.c [hunk 1]
```diff
@@ -319,100 +329,93 @@ process_module_stringtable (MonoObject *filever, gpointer data, guchar lang_hi,
...
+	if (mono_w32process_ver_query_value (data, query, (gpointer *)&trans_data, &trans_size)) {
+		/* use the first language ID we see */
+		if (trans_size >= 4) {
+ 			LOGDEBUG (g_message("%s: %s has 0x%0x 0x%0x 0x%0x 0x%0x", __func__, g_utf16_to_utf8 (filename, -1, NULL, NULL, NULL), trans_data[0], trans_data[1], trans_data[2], trans_data[3]));
+			lang = (trans_data[0]) | (trans_data[1] << 8) | (trans_data[2] << 16) | (trans_data[3] << 24);
+			/* Only give the lower 16 bits to mono_w32process_ver_language_name, as Windows gets confused otherwise  */
+			lang_count = mono_w32process_ver_language_name (lang & 0xFFFF, lang_buf, 128);
+			if (lang_count) {
+				process_set_field_string (filever, "language", lang_buf, lang_count, error);
+				return_if_nok (error);
```
Comment (alexischr) [inline mono/metadata/w32process.c]: Each of the `return_if_nok`s leaks `data` and `query` (afaict the issue was already there)

File: mono/metadata/w32process.c [hunk 2]
```diff
@@ -319,100 +329,93 @@ process_module_stringtable (MonoObject *filever, gpointer data, guchar lang_hi,
...
 			}
+			process_module_stringtable (filever, data, trans_data[0], trans_data[1], error);
+			return_if_nok (error);
```
Comment (alexischr) [inline mono/metadata/w32process.c]: ^

File: mono/metadata/w32process.c [hunk 3]
```diff
@@ -319,100 +329,93 @@ process_module_stringtable (MonoObject *filever, gpointer data, guchar lang_hi,
...
+		}
+	} else {
+		int i;
 
-			g_free (query);
+		for (i = 0; i < G_N_ELEMENTS (stringtable_entries); ++i) {
+			/* No strings, so set every field to the empty string */
+			process_set_field_string (filever, stringtable_entries [i].name, EMPTY_STRING, 0, error);
+			return_if_nok (error);
```
Comment (alexischr) [inline mono/metadata/w32process.c]: ^

File: mono/metadata/w32process.c [hunk 4]
```diff
@@ -319,100 +329,93 @@ process_module_stringtable (MonoObject *filever, gpointer data, guchar lang_hi,
...
+		}
+
+		/* And language seems to be set to en_US according to bug 374600 */
+		lang_count = mono_w32process_ver_language_name (0x0409, lang_buf, 128);
+		if (lang_count) {
+			process_set_field_string (filever, "language", lang_buf, lang_count, error);
+			return_if_nok (error);
```
Comment (alexischr) [inline mono/metadata/w32process.c]: ^

**Comment by luhenry** [comment]
build

**Comment by alexischr** [comment]
Test failure doesn't seem to be related.

**Comment by mfilippov** [comment]
@luhenry Could you backport this patch into 5.8?

**Pull request merged**